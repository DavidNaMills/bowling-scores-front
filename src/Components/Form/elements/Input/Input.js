import React from 'react';
import classes from './Input.module.scss';
import { CreateSharp, ErrorSharp } from '@material-ui/icons'

const iconOk = [classes.icon, classes.icon__required];
const iconErr = [classes.icon, classes.icon__error];

const Input = ({ changed, id, type = 'text', label = null, placeholder = '', error = [], isValid = false, isRequired = false }) => {
    let styles = [classes.input, classes.input__default]

    if (isValid) {
        styles = styles.concat(classes.input__valid);
    }

    if (error.length > 0) {
        styles = styles.concat(classes.input__error);
    }

    return (
        <div className={classes.input__label}>
            {label && <label htmlFor={id}>{label}</label>}
            <div>
                <input
                    className={styles.join(' ')}
                    id={id}
                    type={type}
                    onChange={e => changed(e, id)}
                    placeholder={placeholder}
                />

                {
                    error.length > 0
                        ? <ErrorSharp className={iconErr.join(' ')} />
                        : isRequired && <CreateSharp className={iconOk.join(' ')} />
                }
            </div>
            {error.length > 0 &&
                <div className={classes.input__errorContainer}>
                    {
                        error.map((x, i) => <p key={i} className={classes.input__errorContainer_errorMsg}>{x}</p>, 0)
                    }
                </div>
            }
        </div>
    )
}

export default Input;