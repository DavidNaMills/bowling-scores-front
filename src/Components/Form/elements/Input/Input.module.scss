@import '../../../../styles/colors.scss';
@import '../../../../styles/fonts.scss';

.input{
    border: none;
    padding: 5px;
    width: $input_width;
    
    &__label{
        margin-top: $input_spacing;
    }

    &__default{
        background: rgb(230, 230, 230);
    }

    &__valid{
        -webkit-box-shadow: $valid_input_shadow;
        -moz-box-shadow: $valid_input_shadow;
        box-shadow: $valid_input_shadow;
        border-bottom: solid 1px $btn_default_bkg;
    };
    
    &__error{
        -webkit-box-shadow: $error_input_shadow;
        -moz-box-shadow: $error_input_shadow;
        box-shadow: $error_input_shadow;
        border-bottom: solid 1px $btn_danger_bkg;
    }

    &__errorContainer{
        width: $input_width;
        padding: 5px;

        -webkit-box-shadow: $error_input_shadow;
        -moz-box-shadow: $error_input_shadow;
        box-shadow: $error_input_shadow;

        &_errorMsg{
            font-size: $error_size_small;
            color: $btn_danger_bkg;
        }
    }
}

.icon{
    margin-left: 5px;
    display: flex;
    justify-content: center;
    align-content: center;

    &__required{
        color: $btn_default_bkg_active_hover;
    }

    &__error{
        color: $btn_danger_bkg_active;
    }
}