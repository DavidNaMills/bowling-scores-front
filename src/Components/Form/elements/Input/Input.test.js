import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { CreateSharp, ErrorSharp } from '@material-ui/icons'

import Input from './Input';

describe('<Input/> test suite', () => {
    const testId = 'test';
    const action = { target: { value: 'This is a test', id: testId } }
    const mockFn = jest.fn();
    const testPlaceholder = 'placeholder';
    const testLabel = 'Test Label';
    const errors = ['error 1', 'error 2']
    const base = {
        id: testId,
        changed: mockFn,
        placeholder: testPlaceholder,
        label: testLabel,
        type: 'text'
    }

    describe('Snapshot tests', () => {
        const testData = [
            { 
                snap: 'minimal config snapshot test',
                config: {...base, placeholder:null, label:null}
            },
            { 
                snap: 'basic with placeholder and label snapshot test',
                config: {...base}
            },
            { 
                snap: 'required with label and placeholder snapshot test',
                config: {...base, isRequired: true}
            },
            { 
                snap: 'invalid snapshot test',
                config: {...base, error: errors}
            }
        ]

        testData.forEach(x=>{
            test(x.snap, ()=>{
                const component = renderer.create(<Input {...x.config}/>);
                const app = component.toJSON();
                expect(app).toMatchSnapshot();
            })
        });
    })

    describe('Styling tests', () => {
        it('creates a basic text input: only default values', () => {
            const wrapper = shallow(<Input changed={mockFn} id={testId} />)
            const elem = wrapper.find('input');

            expect(elem.length).toBe(1);
            expect(elem.props().type).toEqual('text');
            expect(elem.props().placeholder).toEqual('');
            expect(elem.hasClass('input__default')).toBeTruthy();

            expect(elem.find('label').length).toBe(0);
        });


        it('creates a password text input: displays label and placeholder', () => {
            const wrapper = shallow(<Input {...base} type='password' />)
            const elem = wrapper.find('input');
            const lbl = wrapper.find('label');

            expect(elem.length).toBe(1);
            expect(elem.props().type).toEqual('password');
            expect(elem.props().placeholder).toEqual(testPlaceholder);
            expect(elem.hasClass('input__default')).toBeTruthy();

            expect(lbl.length).toBe(1);
            expect(lbl.text()).toEqual(testLabel);
            expect(lbl.props().htmlFor).toEqual(testId);
        });

        it('creates an input displays a required icon', () => {
            const wrapper = shallow(<Input {...base} isRequired isValid />)
            const elem = wrapper.find(CreateSharp);
            const inp = wrapper.find('input')
            expect(elem.length).toBe(1);
            expect(elem.hasClass('icon__required')).toBeTruthy();
            expect(inp.hasClass('input__valid')).toBeTruthy();
        })

        it('creates an input that displays an error icon. not required', () => {
            const wrapper = shallow(<Input {...base} error={errors} />)
            const elem = wrapper.find(ErrorSharp);
            const inp = wrapper.find('input');
            const errP = wrapper.find('p');

            expect(elem.length).toBe(1);
            expect(elem.hasClass('icon__error')).toBeTruthy();
            expect(inp.hasClass('input__error')).toBeTruthy();

            expect(errP.length).toBe(errors.length)
            expect(errP.first().hasClass('input__errorContainer_errorMsg')).toBeTruthy();

        });

        it('creates an input that displays an error icon. when required', () => {
            const wrapper = shallow(<Input {...base} error={errors} isRequired />)
            const elem = wrapper.find(ErrorSharp);
            const inp = wrapper.find('input');
            const errP = wrapper.find('p');

            expect(elem.length).toBe(1);
            expect(elem.hasClass('icon__error')).toBeTruthy();
            expect(inp.hasClass('input__error')).toBeTruthy();
            expect(errP.length).toBe(errors.length)
            expect(errP.first().hasClass('input__errorContainer_errorMsg')).toBeTruthy();
        });
    })

    describe('Behaviour tests', () => {
        it('should invoke the action when clicked', () => {
            const wrapper = shallow(<Input {...base} />)
            wrapper.find('input').simulate('change', action);
            expect(mockFn.mock.calls.length).toBe(1);

            //TODO: test correct values are returned

        })
    })
})