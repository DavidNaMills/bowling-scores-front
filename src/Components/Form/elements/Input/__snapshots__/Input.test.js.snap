// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> test suite Snapshot tests basic with placeholder and label snapshot test 1`] = `
<div
  className="input__label"
>
  <label
    htmlFor="test"
  >
    Test Label
  </label>
  <div>
    <input
      className="input input__default"
      id="test"
      onChange={[Function]}
      placeholder="placeholder"
      type="text"
    />
  </div>
</div>
`;

exports[`<Input/> test suite Snapshot tests invalid snapshot test 1`] = `
<div
  className="input__label"
>
  <label
    htmlFor="test"
  >
    Test Label
  </label>
  <div>
    <input
      className="input input__default input__error"
      id="test"
      onChange={[Function]}
      placeholder="placeholder"
      type="text"
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root icon icon__error"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
      />
    </svg>
  </div>
  <div
    className="input__errorContainer"
  >
    <p
      className="input__errorContainer_errorMsg"
    >
      error 1
    </p>
    <p
      className="input__errorContainer_errorMsg"
    >
      error 2
    </p>
  </div>
</div>
`;

exports[`<Input/> test suite Snapshot tests minimal config snapshot test 1`] = `
<div
  className="input__label"
>
  <div>
    <input
      className="input input__default"
      id="test"
      onChange={[Function]}
      placeholder={null}
      type="text"
    />
  </div>
</div>
`;

exports[`<Input/> test suite Snapshot tests required with label and placeholder snapshot test 1`] = `
<div
  className="input__label"
>
  <label
    htmlFor="test"
  >
    Test Label
  </label>
  <div>
    <input
      className="input input__default"
      id="test"
      onChange={[Function]}
      placeholder="placeholder"
      type="text"
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root icon icon__required"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM21.41 6.34l-3.75-3.75-2.53 2.54 3.75 3.75 2.53-2.54z"
      />
    </svg>
  </div>
</div>
`;
